### PR TITLE
EMTF 2017 PR #5 - DQM

### PR DESCRIPTION
**Pull request by abrinke1**
Part 3 (really) of set of PRs following PR #19741 and #19743
PR #19744 is no longer necessary
PR #19928 (CondDB and O2O) will have to be revisited later
Big emulator PR yet to come

Updated EMTF DQM with several improvements:
  * RPC monitoring plots (using updated unpacker)
  * MPC link monitoring (using updated unpacker)
  * CSC LCT and RPC hit timing plots
  * Many low-level plots moved to subfolders; main plots in main directory

Since the DQM is the only consumer of the unpacked "EMTFHit" and "EMTFTrack" collections, removed "EMTFHit2016" and "EMTFTrack2016" products from the unpacker.

### CODE REVIEW

**Comment by abrinke1** [comment]
Hopefully a fairly straightforward update - and will be very useful for us to have in the online DQM!

Please test :)

**Comment by dmitrijus** [comment]
+1

**Comment by abrinke1** [comment]
Oops ... didn't mean to stop the tests - just found a small typo.  Can we re-start the tests in jenkins?

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

**Comment by abrinke1** [comment]
Is there anything to prevent this PR from going in?

Thanks!
Andrew

**Comment by thomreis** [comment]
DQM (@dmitrijus or @vanbesien) need to sign, yes.

**Comment by rekovic** [comment]
@dmitrijus 
We would like this signed and integrated, so can eventually make a 92x PR, as well.

**Comment by dmitrijus** [comment]
+1

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 1]
```diff
@@ -1,13 +1,16 @@
 #include <string>
 #include <vector>
+#include <iostream>
+#include <map>
 
 #include "DQM/L1TMonitor/interface/L1TStage2EMTF.h"
 
+const double PI = 3.14159265358979323846;
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: hi @abrinke1 - please use M_PI

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 2]
```diff
@@ -189,118 +345,265 @@ void L1TStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+  // Maps CSC station and ring to the monitor element index and uses symmetry of the endcaps    
+  std::map<std::pair<int,int>,int> histIndexCSC = { {{1,4}, 9}, {{1,1}, 8}, {{1,2}, 7}, {{1,3}, 6},
+						    {{2,1}, 5}, {{2,2}, 4}, 
+						    {{3,1}, 3}, {{3,2}, 2},
+						    {{4,1}, 1}, {{4,2}, 0} };
+  
+  // Maps RPC staion and ring to the monitor element index and uses symmetry of the endcaps
+  std::map<std::pair<int, int>, int> histIndexRPC = { {{4,3}, 0}, {{4,2}, 1}, {{3,3}, 2}, {{3,2}, 3}, {{2,2}, 4}, {{1,2}, 5} };
+
+  for (std::vector<l1t::EMTFHit>::const_iterator Hit = HitCollection->begin(); Hit != HitCollection->end(); ++Hit) {
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: please use a range-based loop
for ( const auto & hit: HitCollection) {

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 3]
```diff
@@ -189,118 +345,265 @@ void L1TStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+      // Select well-timed tracks: >= 3 hits, with <= 1 in BX != 0
+      if (numHits < 3) continue;
+      for (int iHit = 0; iHit < numHits; ++iHit) {
+	if (Track->Hits().at(iHit).BX() == 0)
+	  numHitsInTrack_BX0++;
+      }
+      if (numHitsInTrack_BX0 < numHits - 1) continue;
+      
+      for (int iHit = 0; iHit < numHits; ++iHit) {
+ 	l1t::EMTFHit tmpHit = Track->Hits().at(iHit);
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: can be a const & (or even better a range-based loop

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 4]
```diff
@@ -189,118 +345,265 @@ void L1TStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+	int sector       = tmpHit.Sector();
+	int subsector    = tmpHit.Subsector();
+	int cscid_offset = (sector - 1) * 9;
+	int neighbor     = tmpHit.Neighbor();
+	int endcap       = tmpHit.Endcap();	
+	int chamber      = tmpHit.Chamber();
+      	
+	int hist_index = 0; 
+	// Maps CSC BX from -2 to 2 to monitor element cscLCTTIming
+	std::map<int, int> histIndexBX = {{0, 4}, {-1, 0}, {1, 1}, {-2, 2}, {2, 3}};
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: make this a const (also true for the other lookup tables above)

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc [hunk 5]
```diff
@@ -189,118 +345,265 @@ void L1TStage2EMTF::analyze(const edm::Event& e, const edm::EventSetup& c) {
...
+	  }
+	  else {
+	    // Map neighbor chambers to "fake" CSC IDs: 1/3 --> 1, 1/6 --> 2, 1/9 --> 3, 2/3 --> 4, 2/9 --> 5, etc.
+	    int cscid_n = (station == 1 ? (cscid / 3) : (station * 2) + ((cscid - 3) / 6) );
+	    cscLCTTiming[histIndexBX[trackhitBX]]->Fill(cscid_n + cscid_offset, endcap * 5.5);
+	  }
+	
+	  // Fill RPC timing with matched CSC LCTs
+	  if (trackhitBX == 0 && ring == 2) {
+	    for (std::vector<l1t::EMTFHit>::const_iterator Hit = HitCollection->begin(); Hit != HitCollection->end(); ++Hit) {
```
Comment (davidlange6) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: range-based loop

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc (hunk 2 above)
Comment (abrinke1) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: Hi David,
This syntax produces the following error:
error: no matching function for call to 'begin(edm::Handle<std::vector<l1t::EMTFHit> >&)'
for (const auto & Hit: HitCollection)

I realize the current syntax isn't the most concise, but you have any other suggestions?
Best,
Andrew

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc (hunk 1 above)
Comment (abrinke1) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: Done

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc (hunk 4 above)
Comment (abrinke1) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: Done

File: DQM/L1TMonitor/src/L1TStage2EMTF.cc (hunk 3 above)
Comment (abrinke1) [inline DQM/L1TMonitor/src/L1TStage2EMTF.cc]: Done

**Comment by abrinke1** [comment]
@davidlange6 All comments should be addressed in latest commit.

**Comment by abrinke1** [comment]
@code-checks robot, the style issues pointed out are mostly in files that have nothing to do with EMTF, which I would prefer not to touch.

**Comment by davidlange6** [comment]
please-test

**Comment by dmitrijus** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by abrinke1** [comment]
Just updated to CMSSW_9_3_0_pre3 so I could try to test out this RelVal error ("136.788 step3" [1]).  My previous nightly-build release was deprecated, so I couldn't run cmsenv on lxplus.  No other changes to the code, so I don't think there's anything to review - but I guess the tests should be triggered again.

Thanks,
Andrew

[1] https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20006/22275/runTheMatrix-results/matrixTests.log

**Comment by davidlange6** [comment]
please test

**Comment by abrinke1** [comment]
Hi @davidlange6 
I identified the issue from the failing RelVal test [1].  Somehow a line got swapped into the wrong place.  Fixed it and re-ran the failing script successfully [2].  Just pushed to my branch - please test.
Thanks,
Andrew

[1] https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20006/22283/runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

[2] https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20006/22283/runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RAW2DIGI_L1Reco_RECO_SKIM_EI_PAT_ALCA_DQM.py

**Comment by dmitrijus** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**